content/blog/2020-12-14-parallel-Java/index.md

```diff
@@ -0,0 +1,126 @@
+
+### Expressiveness and Coverage
+The authors use DPJ to express all available parallelism for the benchmarks (Merge Sort, Monte Carlo, IDEA, K-Means, and Collision Tree), which shows the expressiveness of DPJ.
+
+![evaluation_table_exp](1.png)
+
+### Perforamnce
+The experiment platform is as follows: a Dell R900 multiprocessor running Red Hat Linux with 24 cores, comprising four six-core Xeon CPU, with a total of 48GB main memory.
+
+![evaluation_figure_perf](2.png)
+
+The figure shows the DPJ version speed up compared with the equivalent sequential version of the program. As we can learn from the figure, the actual speedup is reasonable compared with the ideal performance gain, thus proves that the DPJ adds negligible runtime overhead for achieving determinism.
+
+### Porting Effort
+The following table shows the efforts required for changing the Java source code to the equivalent version in DPJ. Programmer need to modify about 10% of the original code, and most of them are about writing RPL arguments when instantiating types, followed by writing method effect summaries.
+
+![evaluation_table_port](3.png)
+
+## Limitations 
+1. Programmer are required to use the correct semantics to enforce the correctness and determinism of the program. For instance, the commuteswith should be used mostly by library and framework code that is written by experienced programmers and extensively tested.
+
+2. When exploring some kind of parallelism in DPJ, programmers are required to relocate the data layout for achieving better performance. For example, to achieve good cache performance in Barness-Hut, the body code that runs in parallel should be copied to the new destination region at the point of re-insertion.
+
+## Questions
+1. Have you ever written any multi-threaded programs in any programming languages? How did you divide up the workload and enforce the synchronization?
+
+2. Have you ever try multi-threading in Java? What is the key limitation of writing concurrent program in the current Java programming model?
+
+3.  Do you think Java is the best language for multi-thread programming? How about use C++ which offers more low-level control of the memory layout?
+
+4. The authors in this paper provides the new features to help the programmer building the programs running in parallel. But the overall effort of writing, testing, and debugging a program with any parallel programming model is dominated by the time required to understand the parallelism and share patterns, and to debug the parallel code. Thus, probably the program synthesis technique could be our final goal? 

```

Comment:
Program synthesis is an interesting idea here. But maybe you could be a little more specific about what role you imagine synthesis playing? Would it write entire parallel algorithms from scratch? Or would it help in some more limited, specific way?